Convert external int fields to uint32 or int32 as appropriate

smarterclayton opened the pull request:
Downgrades prober to int32 externally (not waiting 40 years for a pod to be ready)

Fuzzer is currently broken (internal int is not a valid fuzz target) which has to be fixed prior to merging this.

Addresses some possible problems from #17095.  Main motivation is that this is required for protobuf support because we can't have ambiguous external types.

smarterclayton committed:
Update conversions to handle primitive conversion

smarterclayton commented:
@lavalamp a custom fuzzer func for *int and **int doesn't seem to be invoked - do you know why a custom fuzzer func on a builtin type wouldn't be invoked?

pmorie commented:
@pweil- who will make the changes in SecurityContextConstraints for this?

pweil- commented:
> @pweil- who will make the changes in SecurityContextConstraints for this?

most likely me

smarterclayton commented:
I'm making all these changes in my own PR

On Tue, Nov 17, 2015 at 8:55 AM, Paul Weil <notifications@github.com> wrote:

> @pweil- <https://github.com/pweil-> who will make the changes in
> SecurityContextConstraints for this?
>
> most likely me
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17353#issuecomment-157375846>
> .
>

smarterclayton commented:
Unless it hasn't been merged, in which case yes, that's someone else's
problem ;)

On Tue, Nov 17, 2015 at 1:33 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> I'm making all these changes in my own PR
>
> On Tue, Nov 17, 2015 at 8:55 AM, Paul Weil <notifications@github.com>
> wrote:
>
>> @pweil- <https://github.com/pweil-> who will make the changes in
>> SecurityContextConstraints for this?
>>
>> most likely me
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/17353#issuecomment-157375846>
>> .
>>
>
>

smarterclayton commented:
@lavalamp disregard the fuzz problem, it's because goreflect returns "int64" for "int" in some contexts and "int" in others.  Still trying to figure out why.

lavalamp commented:
That's weird.

lavalamp commented on pkg/util/util.go:
```diff
@@ -139,7 +139,7 @@ func Until(f func(), period time.Duration, stopCh <-chan struct{}) {
 type IntOrString struct {
 	Kind   IntstrKind
-	IntVal int
+	IntVal int32
```
It is really weird to see code like int(blah.IntVal). I can live with it if I have to, but I'd prefer not to...

* Leave as int, return error if it's > 2B when marshalling? (UGH)
* Change name to Int32Val? 
* Add a `Int() int` function that does the cast for you?

lavalamp commented on pkg/util/util.go:
```diff
@@ -161,6 +161,10 @@ func NewIntOrStringFromString(val string) IntOrString {
 }
 
+func (intstr *IntOrString) IntValue() int {
+	return int(intstr.IntVal)
```
Oh, this function already exists. Can we call it instead of casting?

smarterclayton replied:
Call it everywhere, you mean?  Happy to.

On Tue, Nov 17, 2015 at 4:24 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/util/util.go
> <https://github.com/kubernetes/kubernetes/pull/17353#discussion_r45123105>
> :
>
> > @@ -161,6 +161,10 @@ func NewIntOrStringFromString(val string) IntOrString {
> >  	return IntOrString{Kind: IntstrString, StrVal: val}
> >  }
> >
> > +func (intstr *IntOrString) IntValue() int {
> > +	return int(intstr.IntVal)
>
> Oh, this function already exists. Can we call it instead of casting?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17353/files#r45123105>.
>

smarterclayton commented:
It was a bad fuzzer - we were recreating the fuzzer and it was non obvious
anywhere.

On Tue, Nov 17, 2015 at 4:17 PM, Daniel Smith <notifications@github.com>
wrote:

> That's weird.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17353#issuecomment-157509886>
> .
>

smarterclayton commented:
This is ready for review (still verifying i didn't break any tests).  Goal was 100% no `int` or `uint` in public code, if there are deeper refactorings or other changes I can do those in a follow up.

smarterclayton commented:
I bumped `Prober` fields to uint32 because we shouldn't be waiting multiple years for a pod to become live.

smarterclayton commented:
ok to test

smarterclayton commented:
How can I kick unit/integration?

liggitt commented:
@k8s-bot unit test this

thockin commented on pkg/api/types.go:
```diff
@@ -719,11 +719,11 @@ type Probe struct {
 	Handler `json:",inline"`
 	// Length of time before health checking is activated.  In seconds.
-	InitialDelaySeconds int64 `json:"initialDelaySeconds,omitempty"`
+	InitialDelaySeconds int `json:"initialDelaySeconds,omitempty"`
```
shouldn't we be crystallizing all 'int' to fixed-width?

liggitt replied:
+1

liggitt replied:
Oh, hrmm, this was internal… iirc, the ripple was huge

liggitt replied:
@smarterclayton how many bullets do you feel like biting?

smarterclayton replied:
Internal changes are massive.  If you hate me, make me do it, but in a follow up.

smarterclayton replied:
These are 32 ext and 32 or 64 internal, which is ok for now.

thockin replied:
We should document that in the conventions.  It's obviously easier to use `int` internally.  Will the codegen converters convert int32 -> int ?

smarterclayton replied:
Yes, and generated conversions now properly tolerate the change.

Changing Int internally rippled out to most code paths - if we document
int64 external requires int64 internal (until we nuke internal), we can
attack it in pieces.

On Nov 17, 2015, at 11:29 PM, Tim Hockin <notifications@github.com> wrote:

In pkg/api/types.go
<https://github.com/kubernetes/kubernetes/pull/17353#discussion_r45158886>:

> @@ -719,11 +719,11 @@ type Probe struct {
>  	// The action taken to determine the health of a container
>  	Handler `json:",inline"`
>  	// Length of time before health checking is activated.  In seconds.
> -	InitialDelaySeconds int64 `json:"initialDelaySeconds,omitempty"`
> +	InitialDelaySeconds int `json:"initialDelaySeconds,omitempty"`

We should document that in the conventions. It's obviously easier to use int
internally. Will the codegen converters convert int32 -> int ?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17353/files#r45158886>.

thockin replied:
If teh conversions handle it, fine.  for internal it is a lot less terrible.

On Tue, Nov 17, 2015 at 10:43 PM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/api/types.go
> <https://github.com/kubernetes/kubernetes/pull/17353#discussion_r45164702>
> :
>
> > @@ -719,11 +719,11 @@ type Probe struct {
> >  	// The action taken to determine the health of a container
> >  	Handler `json:",inline"`
> >  	// Length of time before health checking is activated.  In seconds.
> > -	InitialDelaySeconds int64 `json:"initialDelaySeconds,omitempty"`
> > +	InitialDelaySeconds int `json:"initialDelaySeconds,omitempty"`
>
> Yes, and generated conversions now properly tolerate the change. Changing
> Int internally rippled out to most code paths - if we document int64
> external requires int64 internal (until we nuke internal), we can attack it
> in pieces. On Nov 17, 2015, at 11:29 PM, Tim Hockin <
> notifications@github.com> wrote: In pkg/api/types.go <#17353 (comment)
> <https://github.com/kubernetes/kubernetes/pull/17353#discussion_r45158886>
> >:
> @@ -719,11 +719,11 @@ type Probe struct { // The action taken to determine
> the health of a container Handler `json:",inline"` // Length of time before
> health checking is activated. In seconds. - InitialDelaySeconds int64
> `json:"initialDelaySeconds,omitempty"` + InitialDelaySeconds int
> `json:"initialDelaySeconds,omitempty"`
> We should document that in the conventions. It's obviously easier to use
> int internally. Will the codegen converters convert int32 -> int ? — Reply
> to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/17353/files#r45158886>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17353/files#r45164702>.
>

thockin commented on pkg/api/types.go:
```diff
@@ -1051,7 +1051,7 @@ type PodSecurityContext struct {
 	//
 	// If unset, the Kubelet will not modify the ownership and permissions of any volume.
-	FSGroup *int64 `json:"fsGroup,omitempty"`
+	FSGroup *uint32 `json:"fsGroup,omitempty"`
```
@pmorie if we had a type for UnixUID and UnixGID this would be easier... :)

smarterclayton commented:
I can't see what is failing on unit/int - I have green locally

liggitt commented on api/swagger-spec/v1.json:
```diff
@@ -13666,7 +13666,6 @@
      },
      "runAsUser": {
       "type": "integer",
-      "format": "int64",
```
missing format?

smarterclayton replied:
Swagger considers int32 the canonical "integer".

smarterclayton replied:
And there is no support for uint in swagger

liggitt replied:
> Swagger considers int32 the canonical "integer".
> And there is no support for uint in swagger

nice. the integer's integer.

smarterclayton commented:
@k8s-bot unit test this

smarterclayton commented:
@k8s-oncall, is/was the Jenkins unit/integration bot broken?  Haven't been able to have it rerun the unit/integration test?  The normal 'unit test this' isn't working.

gmarek commented:
@k8s-bot unit test this

smarterclayton commented:
Maybe I got reset off the list of people authorized to ask for that?

smarterclayton commented:
ok to test

smarterclayton commented:
ok to test

smarterclayton committed:
Unversioned types should not use ambiguous go-int

All external types that are not int64 are now marked as int32,
including
IntOrString. Prober is now int32 (43 years should be enough of an initial
probe time for anyone).

Did not change the metadata fields for now.

smarterclayton committed:
Generated code

smarterclayton committed:
Fix tests to pass with changed external types

smarterclayton committed:
Generated docs

pmorie commented on pkg/api/errors/errors.go:
```diff
@@ -226,7 +226,7 @@ func NewServerTimeout(kind, operation string, retryAfterSeconds int) error {
 			Kind:              kind,
 			Name:              operation,
-			RetryAfterSeconds: retryAfterSeconds,
+			RetryAfterSeconds: int32(retryAfterSeconds),
```
I am guessing the narrowing/widening conversions here are because you don't want to change the API all over town, or do you really prefer the API for this to be an int.

lavalamp replied:
I don't expect it makes a difference for this field, but casting like this is not safe in general as it potentially hides a cross-platform bug.

smarterclayton replied:
Agree, but I'd rather refactor the internal in a follow on (remove casts to
int32)

On Wed, Nov 18, 2015 at 6:04 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/api/errors/errors.go
> <https://github.com/kubernetes/kubernetes/pull/17353#discussion_r45275000>
> :
>
> > @@ -226,7 +226,7 @@ func NewServerTimeout(kind, operation string, retryAfterSeconds int) error {
> >  		Details: &unversioned.StatusDetails{
> >  			Kind:              kind,
> >  			Name:              operation,
> > -			RetryAfterSeconds: retryAfterSeconds,
> > +			RetryAfterSeconds: int32(retryAfterSeconds),
>
> I don't expect it makes a difference for this field, but casting like this
> is not safe in general as it potentially hides a cross-platform bug.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17353/files#r45275000>.
>

smarterclayton commented:
This is green - generous review appreciated from any and all... :)

lavalamp commented:
Will look shortly

lavalamp commented on contrib/mesos/pkg/service/endpoints_controller.go:
```diff
@@ -429,8 +429,8 @@ func findPort(pod *api.Pod, svcPort *api.ServicePort) (int, int, error) {
 		p := portName.IntVal
```
Call the function instead of casting p below?

lavalamp commented on pkg/api/errors/errors.go:
```diff
@@ -412,7 +412,7 @@ func SuggestsClientDelay(err error) (int, bool) {
 			switch t.Status().Reason {
 			case unversioned.StatusReasonServerTimeout, unversioned.StatusReasonTimeout:
-				return t.Status().Details.RetryAfterSeconds, true
+				return int(t.Status().Details.RetryAfterSeconds), true
```
This direction, casting from int32 to int, should be safe everywhere.

lavalamp commented on pkg/api/testing/fuzzer.go:
```diff
@@ -45,6 +45,17 @@ func FuzzerFor(t *testing.T, version string, src rand.Source) *fuzz.Fuzzer {
+		func(j *int, c fuzz.Continue) {
+			*j = int(c.Int31())
+		},
+		func(j **int, c fuzz.Continue) {
```
I'm surprised you need this function, it should be the default behavior.

smarterclayton replied:
int is 64 bit internally, so fuzzInt generates a 64 bit number which won't
cross the boundary..  That won't round trip to the external types.  If /
when we refactor the internal, we would not need to do this.

On Wed, Nov 18, 2015 at 6:09 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/api/testing/fuzzer.go
> <https://github.com/kubernetes/kubernetes/pull/17353#discussion_r45275625>
> :
>
> > @@ -45,6 +45,17 @@ func FuzzerFor(t *testing.T, version string, src rand.Source) *fuzz.Fuzzer {
> >  		f.RandSource(src)
> >  	}
> >  	f.Funcs(
> > +		func(j *int, c fuzz.Continue) {
> > +			*j = int(c.Int31())
> > +		},
> > +		func(j **int, c fuzz.Continue) {
>
> I'm surprised you need this function, it should be the default behavior.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17353/files#r45275625>.
>

lavalamp replied:
Ah, I see.

lavalamp commented on pkg/api/testing/fuzzer.go:
```diff
@@ -138,8 +149,8 @@ func FuzzerFor(t *testing.T, version string, src rand.Source) *fuzz.Fuzzer {
 			c.FuzzNoCustom(j) // fuzz self without calling this function again
-			completions := c.Rand.Int()
-			parallelism := c.Rand.Int()
+			completions := int(c.Rand.Int31())
```
The versioned struct is int32 and the internal is int?

smarterclayton replied:
Correct

On Wed, Nov 18, 2015 at 6:20 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/api/testing/fuzzer.go
> <https://github.com/kubernetes/kubernetes/pull/17353#discussion_r45276792>
> :
>
> > @@ -138,8 +149,8 @@ func FuzzerFor(t *testing.T, version string, src rand.Source) *fuzz.Fuzzer {
> >  		},
> >  		func(j *extensions.JobSpec, c fuzz.Continue) {
> >  			c.FuzzNoCustom(j) // fuzz self without calling this function again
> > -			completions := c.Rand.Int()
> > -			parallelism := c.Rand.Int()
> > +			completions := int(c.Rand.Int31())
>
> The versioned struct is int32 and the internal is int?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17353/files#r45276792>.
>

lavalamp commented on pkg/api/testing/fuzzer.go:
```diff
@@ -361,9 +371,9 @@ func FuzzerFor(t *testing.T, version string, src rand.Source) *fuzz.Fuzzer {
+			minReplicas := int(c.Rand.Int31())
 			s.MinReplicas = &minReplicas
-			s.CPUUtilization = &extensions.CPUTargetUtilization{TargetPercentage: int(c.RandUint64())}
+			s.CPUUtilization = &extensions.CPUTargetUtilization{TargetPercentage: int(c.Int31())}
```
This will no longer test negative numbers.

smarterclayton replied:
Going to fix with int(int32(c.RandUint64()))

On Wed, Nov 18, 2015 at 6:22 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/api/testing/fuzzer.go
> <https://github.com/kubernetes/kubernetes/pull/17353#discussion_r45276960>
> :
>
> >  			s.MinReplicas = &minReplicas
> > -			s.CPUUtilization = &extensions.CPUTargetUtilization{TargetPercentage: int(c.RandUint64())}
> > +			s.CPUUtilization = &extensions.CPUTargetUtilization{TargetPercentage: int(c.Int31())}
>
> This will no longer test negative numbers.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17353/files#r45276960>.
>

smarterclayton committed:
Review comments

lavalamp commented on pkg/apis/extensions/v1beta1/defaults_test.go:
```diff
@@ -294,9 +294,9 @@ func roundTrip(t *testing.T, obj runtime.Object) runtime.Object {
-func newInt(val int) *int {
-	p := new(int)
-	*p = val
+func newInt(val int) *int32 {
```
Fix function name?

lavalamp commented on pkg/util/intstr/intstr.go:
```diff
@@ -44,7 +44,7 @@ const (
 
 // FromInt creates an IntOrString object with an int value.
```
Add: it's your responsibility to not call this with an int that won't fit in 32 bits.

lavalamp commented on pkg/util/intstr/intstr.go:
```diff
@@ -67,7 +67,17 @@ func (intstr *IntOrString) String() string {
+	return strconv.Itoa(intstr.IntValue())
+}
+
+// IntValue returns the IntVal cast as int32 if type Int, or if
```
s/32//

lavalamp commented:
minor comments

I don't like that the conversion functions fail silently on ints bigger than 32 bits, but I was waiting to see a field for which it would matter to comment about it and I didn't find one. But this int/int32 casting business is a bad code smell.

LGTM

lavalamp commented:
If this wasn't such an extensive change, I'd ask for a squash. :)

smarterclayton commented:
Trivial rebase, reapplying LGTM

wojtek-t commented:
@smarterclayton - can you please rebase this PR?

smarterclayton commented:
Trivial rebase, reapplying LGTM

wojtek-t commented:
hmm - not sure what exactly happened, but it needs rebase again

smarterclayton commented:
possible e2e flake on `create a job from an image when restart is OnFailure`

smarterclayton commented:
ok to test

smarterclayton commented:
Trivial rebase, reapplying LGTM

thockin commented on pkg/util/intstr/intstr.go:
```diff
@@ -67,7 +71,17 @@ func (intstr *IntOrString) String() string {
+// IntValue returns the IntVal if type Int, or if
+// it is a String, will attempt a conversion to int.
+func (intstr *IntOrString) IntValue() int {
+	if intstr.Type == String {
```
I think this is wrong and we should not do it.  It's scope creep for this PR anyway.  Was there a use case justifying this?  I think we should roll it back

smarterclayton replied:
It was to prevent having to refactor 900 lines of internal code that assume
IntOrString is int, not int32.  The string conversion bit can be removed as
long as we doc it (all the places it was used was the IntValue)

On Mon, Nov 30, 2015 at 1:27 PM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/util/intstr/intstr.go
> <https://github.com/kubernetes/kubernetes/pull/17353#discussion_r46181719>
> :
>
> > @@ -67,7 +71,17 @@ func (intstr *IntOrString) String() string {
> >  	if intstr.Type == String {
> >  		return intstr.StrVal
> >  	}
> > -	return strconv.Itoa(intstr.IntVal)
> > +	return strconv.Itoa(intstr.IntValue())
> > +}
> > +
> > +// IntValue returns the IntVal if type Int, or if
> > +// it is a String, will attempt a conversion to int.
> > +func (intstr *IntOrString) IntValue() int {
> > +	if intstr.Type == String {
>
> I think this is wrong and we should not do it. It's scope creep for this
> PR anyway. Was there a use case justifying this? I think we should roll it
> back
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17353/files#r46181719>.
>

thockin replied:
I don't mean the function, I mean the new cast of string -> int.  I assert
that this is a discriminated union, and not a "best effort to figure out
what the user meant".  if they said type String and read it back as an int,
they get a 0 or an error or a panic.  Not an atoi() of the string.

On Mon, Nov 30, 2015 at 11:44 AM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/util/intstr/intstr.go
> <https://github.com/kubernetes/kubernetes/pull/17353#discussion_r46192602>
> :
>
> > @@ -67,7 +71,17 @@ func (intstr *IntOrString) String() string {
> >  	if intstr.Type == String {
> >  		return intstr.StrVal
> >  	}
> > -	return strconv.Itoa(intstr.IntVal)
> > +	return strconv.Itoa(intstr.IntValue())
> > +}
> > +
> > +// IntValue returns the IntVal if type Int, or if
> > +// it is a String, will attempt a conversion to int.
> > +func (intstr *IntOrString) IntValue() int {
> > +	if intstr.Type == String {
>
> It was to prevent having to refactor 900 lines of internal code that
> assume IntOrString is int, not int32. The string conversion bit can be
> removed as long as we doc it (all the places it was used was the IntValue)
> … <#15159ed0968e77c8_>
> On Mon, Nov 30, 2015 at 1:27 PM, Tim Hockin <notifications@github.com>
> wrote: In pkg/util/intstr/intstr.go <#17353 (comment)
> <https://github.com/kubernetes/kubernetes/pull/17353#discussion_r46181719>>
> : > @@ -67,7 +71,17 @@ func (intstr *IntOrString) String() string { > if
> intstr.Type == String { > return intstr.StrVal > } > - return
> strconv.Itoa(intstr.IntVal) > + return strconv.Itoa(intstr.IntValue()) > +}
> > + > +// IntValue returns the IntVal if type Int, or if > +// it is a
> String, will attempt a conversion to int. > +func (intstr *IntOrString)
> IntValue() int { > + if intstr.Type == String { I think this is wrong and
> we should not do it. It's scope creep for this PR anyway. Was there a use
> case justifying this? I think we should roll it back — Reply to this email
> directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/17353/files#r46181719>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17353/files#r46192602>.
>

smarterclayton replied:
I will nuke the best effort in a pr.

On Dec 1, 2015, at 1:38 AM, Tim Hockin <notifications@github.com> wrote:

In pkg/util/intstr/intstr.go
<https://github.com/kubernetes/kubernetes/pull/17353#discussion_r46244484>:

> @@ -67,7 +71,17 @@ func (intstr *IntOrString) String() string {
>  	if intstr.Type == String {
>  		return intstr.StrVal
>  	}
> -	return strconv.Itoa(intstr.IntVal)
> +	return strconv.Itoa(intstr.IntValue())
> +}
> +
> +// IntValue returns the IntVal if type Int, or if
> +// it is a String, will attempt a conversion to int.
> +func (intstr *IntOrString) IntValue() int {
> +	if intstr.Type == String {

I don't mean the function, I mean the new cast of string -> int. I assert
that this is a discriminated union, and not a "best effort to figure out
what the user meant". if they said type String and read it back as an int,
they get a 0 or an error or a panic. Not an atoi() of the string.
… <#>
On Mon, Nov 30, 2015 at 11:44 AM, Clayton Coleman <notifications@github.com>
wrote: In pkg/util/intstr/intstr.go <#17353 (comment)
<https://github.com/kubernetes/kubernetes/pull/17353#discussion_r46192602>>
: > @@ -67,7 +71,17 @@ func (intstr *IntOrString) String() string { > if
intstr.Type == String { > return intstr.StrVal > } > - return
strconv.Itoa(intstr.IntVal) > + return strconv.Itoa(intstr.IntValue()) > +}
> + > +// IntValue returns the IntVal if type Int, or if > +// it is a
String, will attempt a conversion to int. > +func (intstr *IntOrString)
IntValue() int { > + if intstr.Type == String { It was to prevent having to
refactor 900 lines of internal code that assume IntOrString is int, not
int32. The string conversion bit can be removed as long as we doc it (all
the places it was used was the IntValue) … <#15159ed0968e77c8_> On Mon, Nov
30, 2015 at 1:27 PM, Tim Hockin <notifications@github.com> wrote: In
pkg/util/intstr/intstr.go <#17353
<https://github.com/kubernetes/kubernetes/pull/17353> (comment) <#17353
(comment)
<https://github.com/kubernetes/kubernetes/pull/17353#discussion_r46181719>>>
: > @@ -67,7 +71,17 @@ func (intstr *IntOrString) String() string { > if
intstr.Type == String { > return intstr.StrVal > } > - return
strconv.Itoa(intstr.IntVal) > + return strconv.Itoa(intstr.IntValue()) > +}
> + > +// IntValue returns the IntVal if type Int, or if > +// it is a
String, will attempt a conversion to int. > +func (intstr *IntOrString)
IntValue() int { > + if intstr.Type == String { I think this is wrong and
we should not do it. It's scope creep for this PR anyway. Was there a use
case justifying this? I think we should roll it back — Reply to this email
directly or view it on GitHub <
https://github.com/kubernetes/kubernetes/pull/17353/files#r46181719>. —
Reply to this email directly or view it on GitHub <
https://github.com/kubernetes/kubernetes/pull/17353/files#r46192602>.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17353/files#r46244484>.

nikhiljindal commented:
@smarterclayton can you please add this to https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md as well?

smarterclayton commented:
Weird, I thought we added it.  Yes

On Thu, Dec 17, 2015 at 2:19 PM, Nikhil Jindal <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> can you please add
> this to
> https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md
> as well?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17353#issuecomment-165553858>
> .
>

smarterclayton commented:
Opened https://github.com/kubernetes/kubernetes/pull/18941

On Thu, Dec 17, 2015 at 3:05 PM, Clayton Coleman <ccoleman@redhat.com> wrote:
> Weird, I thought we added it.  Yes
>
> On Thu, Dec 17, 2015 at 2:19 PM, Nikhil Jindal <notifications@github.com>
> wrote:
>>
>> @smarterclayton can you please add this to
>> https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md
>> as well?
>>
>> —
>> Reply to this email directly or view it on GitHub.
>
>
